File: 2021/days/11/day11.js
const { parseTextFile, logger } = require("../../helpers");
const isNumber = 0;
let solution = 0;
let octopus = [];
function part1(isTest) {
  logger(
    `\n---------- Day 11 part 1 ${isTest ? "(TESTINPUT)" : ""} ----------`,
    "cyan"
  );
  const input = parseTextFile(__dirname, isTest, isNumber, "\n");
  let steps = 10;
  solution = 0;
  octopus = [];
  input.forEach((row) => {
    octopus.push(row.split("").map(Number));
  });

  for (let index = 1; index <= steps; index++) {
    console.log("Step:", index);
    for (let y = 0; y < octopus.length; y++) {
      for (let x = 0; x < octopus[y].length; x++) {
        octopus[y][x]++;
      }
    }
    for (let y = 0; y < octopus.length; y++) {
      for (let x = 0; x < octopus[y].length; x++) {
        if (octopus[y][x] === 10) {
          flash(y, x);
        }
      }
    }
    // Reset flashed octos
    for (let y = 0; y < octopus.length; y++) {
      for (let x = 0; x < octopus[y].length; x++) {
        if (octopus[y][x] === -1) {
          octopus[y][x] = 0;
        }
      }
    }
  }
  console.log("Answer:", solution);
}

function part2(isTest) {
  logger(
    `\n---------- Day 11 part 2 ${isTest ? "(TESTINPUT)" : ""} ----------`,
    "cyan"
  );
  const input = parseTextFile(__dirname, isTest, isNumber, "\n");
  let steps = 10;
  solution = 0;
  octopus = [];
  input.forEach((row) => {
    octopus.push(row.split("").map(Number));
  });
  console.log("Answer:", solution);
}

function flash(y, x) {
  solution++;
  octopus[y][x] = -1;
  // Locate neighbours
  console.log(`octopus[${y}][${x}]`);
  let adj = [-1, 0, 1];
  for (row of adj) {
    for (col of adj) {
      let yy = y + row;
      let xx = x + col;
      if (
        yy >= 0 &&
        yy < octopus.length &&
        xx >= 0 &&
        xx < octopus[0].length &&
        octopus[yy][xx] !== -1 // Handle the octopus[y][x] itself
      ) {
        // Increment the number and check if it needs to flash
        octopus[yy][xx]++;
        if (octopus[yy][xx] >= 10) {
          flash(yy, xx);
        }
      }
    }
  }
}

module.exports = {
  part1,
  part2,
};

File: 2021/days/11/day11.py
test = "test.txt"
infile = "input.txt"
isTest = 0
if(isTest):
    infile = test

octopus = []
for line in open(infile):
    octopus.append([int(x) for x in line.strip()])


rowLength = len(octopus)
columnLength = len(octopus[0])

solution = 0
steps = 0


def flash(y,x):
    global solution
    solution += 1
    octopus[y][x] = -1
    neighbours = [-1,0,1]
    for row in neighbours:
        for col in neighbours:
            yy = y+row
            xx = x+col
            if 0<=yy<rowLength and 0<=xx<columnLength and octopus[yy][xx]!=-1:
                octopus[yy][xx] += 1
                if octopus[yy][xx] >= 10:
                    flash(yy,xx)

while True:
    steps += 1
    for y in range(rowLength):
        for x in range(columnLength):
            octopus[y][x] += 1
    for y in range(rowLength):
        for x in range(columnLength):
            if octopus[y][x] == 10:
                flash(y,x)
    allFlashed = True
    for y in range(rowLength):
        for x in range(columnLength):
            if octopus[y][x] == -1:
                octopus[y][x] = 0
            else:
                allFlashed = False
    if steps == 100:
        print("Part 1:", solution)
        # break 
    if allFlashed:
        print("Part 2:", steps)
        break





